### PR TITLE
Update CHANGELOG for 1.12.2

### PR DESCRIPTION
**Pull request by vieux**
Signed-off-by: Victor Vieux vieux@docker.com

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,43 @@ information on the list of deprecated flags and APIs please have a look at
 https://docs.docker.com/engine/deprecated/ where target removal dates can also
 be found.
 
+## 1.12.2
+
+### Runtime
+
+* Check for non-nil container after match [#26049](https://github.com/docker/docker/pull/26049)
```
Comment (thaJeztah) [inline CHANGELOG.md]: Generally, we try to describe what is being fixed (not how). Perhaps something like;

```Markdown
Fix a panic due to a race condition filtering `docker ps` 
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -5,6 +5,43 @@ information on the list of deprecated flags and APIs please have a look at
...
+* Add retry logic during aufs unmount [#26536](https://github.com/docker/docker/pull/26536)
```
Comment (thaJeztah) [inline CHANGELOG.md]:
```markdown
Implement retry logic to prevent "Unable to remove filesystem" errors when using the aufs storage driver
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -5,6 +5,43 @@ information on the list of deprecated flags and APIs please have a look at
...
+* Prevent libdevmapper from deleting /dev/mapper/docker-xxx symbolic links in RemoveDeviceDeferred [#24740](https://github.com/docker/docker/pull/24740)
```
Comment (thaJeztah) [inline CHANGELOG.md]:
```Markdown
Prevent devicemapper from removing device symlinks if `dm.use_deferred_removal` is enabled.
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -5,6 +5,43 @@ information on the list of deprecated flags and APIs please have a look at
...
+* Fixing Regression in exit codes [#26777](https://github.com/docker/docker/pull/26777)
```
Comment (thaJeztah) [inline CHANGELOG.md]:
```Markdown
Fix an issue where the CLI did not return correct exit codes if a command was run with invalid options.
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -5,6 +5,43 @@ information on the list of deprecated flags and APIs please have a look at
...
+* Prevent stdout / stderr race condition in limitedBuffer [#26507](https://github.com/docker/docker/pull/26507)
```
Comment (thaJeztah) [inline CHANGELOG.md]:
```Markdown
Fix a panic due to a bug in stdout / stderr processing in health checks.
```

File: CHANGELOG.md [hunk 6]
```diff
@@ -5,6 +5,43 @@ information on the list of deprecated flags and APIs please have a look at
...
+* Fix exec's children handling [#26874](https://github.com/docker/docker/pull/26874)
```
Comment (thaJeztah) [inline CHANGELOG.md]: Is this related to https://github.com/docker/docker/issues/25348 ?

File: CHANGELOG.md [hunk 7]
```diff
@@ -5,6 +5,43 @@ information on the list of deprecated flags and APIs please have a look at
...
+
+### Swarm Mode
+
+* Do not autodetect advertise address on join [#26211](https://github.com/docker/docker/pull/26211)
+* Vendor SwarmKit [#26765](https://github.com/docker/docker/pull/26765)
+ * Bounce session after failed status update [docker/swarmkit#1539](https://github.com/docker/swarmkit/pull/1539)
+ * Fix possible deadlocks [docker/swarmkit#1537](https://github.com/docker/swarmkit/pull/1537)
+ * Update boltdb dependency [docker/swarmkit#1510](https://github.com/docker/swarmkit/pull/1510)
+ * Fix panic in portallocator and deallocate endpoints when update with none EndpointSpec [docker/swarmkit#1481](https://github.com/docker/swarmkit/pull/1481)
+ * Validate duplicate published port in ServiceSpec [docker/swarmkit#1495](https://github.com/docker/swarmkit/pull/1495)
```
Comment (thaJeztah) [inline CHANGELOG.md]: Perhaps:

```Markdown
Produce an error if the same port is published twice on `service create` or `service update`.
```

File: CHANGELOG.md [hunk 8]
```diff
@@ -5,6 +5,43 @@ information on the list of deprecated flags and APIs please have a look at
...
+ * Check entire service spec when deciding whether to replace ongoing update [docker/swarmkit#1497](https://github.com/docker/swarmkit/pull/1497)
```
Comment (thaJeztah) [inline CHANGELOG.md]:
```Markdown
Fix an issue where changes to a service were not detected, resulting in the service not being updated.
```

File: CHANGELOG.md [hunk 9]
```diff
@@ -5,6 +5,43 @@ information on the list of deprecated flags and APIs please have a look at
...
+
+### Contrib
+
+* Debian sysvinitscript one missed DOCKER->DOCKERD change [#25869](https://github.com/docker/docker/pull/25869)
```
Comment (thaJeztah) [inline CHANGELOG.md]:
```Markdown
Update the debian sysv-init script to use `dockerd` instead of `docker daemon`
```

File: CHANGELOG.md [hunk 10]
```diff
@@ -5,6 +5,43 @@ information on the list of deprecated flags and APIs please have a look at
...
+* Patch Go with fix for OSX Sierra [#26875](https://github.com/docker/docker/pull/26875)
```
Comment (thaJeztah) [inline CHANGELOG.md]:
```Markdown
Improve stability when running the docker client on MacOS Sierra.
```

File: CHANGELOG.md [hunk 11]
```diff
@@ -5,6 +5,43 @@ information on the list of deprecated flags and APIs please have a look at
...
+* Remove go SHA check in ppc64le [#24335](https://github.com/docker/docker/pull/24335)
```
Comment (thaJeztah) [inline CHANGELOG.md]: Not sure this is important enough to mention in the changelog (afaik only IBM builds for ppc64le, and they are aware of this change)

File: CHANGELOG.md [hunk 12]
```diff
@@ -5,6 +5,43 @@ information on the list of deprecated flags and APIs please have a look at
...
+
+### Windows
+
+* Force input stream ANSI emulation for ConEmu [#25578](https://github.com/docker/docker/pull/25578)
```
Comment (thaJeztah) [inline CHANGELOG.md]:
```Markdown
Fix an issue where arrow-navigation did not work when running the docker client in ConEmu
```

File: CHANGELOG.md [hunk 13]
```diff
@@ -5,6 +5,43 @@ information on the list of deprecated flags and APIs please have a look at
 https://docs.docker.com/engine/deprecated/ where target removal dates can also
 be found.
 
+## 1.12.2
+
```
Comment (thaJeztah) [inline CHANGELOG.md]: Can you repeat the whole `IMPORTANT: Docker 1.12 ships with an updated systemd  ... ` section for this release? Not everyone reads older release notes, so perhaps better "safe" than "sorry" for those that do not check older versions :smile:

File: CHANGELOG.md [hunk 14]
```diff
@@ -5,6 +5,43 @@ information on the list of deprecated flags and APIs please have a look at
 https://docs.docker.com/engine/deprecated/ where target removal dates can also
 be found.
 
+## 1.12.2
```
Comment (thaJeztah) [inline CHANGELOG.md]: 1.12.2-rc2 (2016-09-27) ?

(Don't know, I see previous RC's also omitted `-rcX`)

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Left some suggestions, but there's probably more :smile:

**Comment by mlaventure** [comment]
`-` should be used for fix, `*` for upgrade (`+` for new features, but they shouldn't any in that release)

**Comment by thaJeztah** [comment]
ping @aaronlehmann @aboch @mrjana PTAL!

File: CHANGELOG.md (hunk 6 above)
Comment (vieux) [inline CHANGELOG.md]: idk, @mlaventure ?

File: CHANGELOG.md [hunk 15]
```diff
@@ -5,6 +5,43 @@ information on the list of deprecated flags and APIs please have a look at
...
+ * Avoid returning early on agent join failures [docker/libnetwork#1473](https://github.com/docker/libnetwork/pull/1473)
+
+### Swarm Mode
+
+* Do not autodetect advertise address on join [#26211](https://github.com/docker/docker/pull/26211)
+* Vendor SwarmKit [#26765](https://github.com/docker/docker/pull/26765)
+ * Bounce session after failed status update [docker/swarmkit#1539](https://github.com/docker/swarmkit/pull/1539)
+ * Fix possible deadlocks [docker/swarmkit#1537](https://github.com/docker/swarmkit/pull/1537)
+ * Update boltdb dependency [docker/swarmkit#1510](https://github.com/docker/swarmkit/pull/1510)
+ * Fix panic in portallocator and deallocate endpoints when update with none EndpointSpec [docker/swarmkit#1481](https://github.com/docker/swarmkit/pull/1481)
```
Comment (aaronlehmann) [inline CHANGELOG.md]: "Fix panic and endpoint leak when a service is updated with no endpoints"

File: CHANGELOG.md [hunk 16]
```diff
@@ -5,6 +5,43 @@ information on the list of deprecated flags and APIs please have a look at
...
+* Vendor libnetwork [#26879](https://github.com/docker/docker/pull/26879) [#26953](https://github.com/docker/docker/pull/26953)
+ * Avoid returning early on agent join failures [docker/libnetwork#1473](https://github.com/docker/libnetwork/pull/1473)
+
+### Swarm Mode
+
+* Do not autodetect advertise address on join [#26211](https://github.com/docker/docker/pull/26211)
+* Vendor SwarmKit [#26765](https://github.com/docker/docker/pull/26765)
+ * Bounce session after failed status update [docker/swarmkit#1539](https://github.com/docker/swarmkit/pull/1539)
+ * Fix possible deadlocks [docker/swarmkit#1537](https://github.com/docker/swarmkit/pull/1537)
+ * Update boltdb dependency [docker/swarmkit#1510](https://github.com/docker/swarmkit/pull/1510)
```
Comment (aaronlehmann) [inline CHANGELOG.md]: This isn't technically related to revendoring swarmkit. There was an engine PR that did it as well.

File: CHANGELOG.md [hunk 17]
```diff
@@ -5,6 +5,43 @@ information on the list of deprecated flags and APIs please have a look at
...
+### Networking
+* Vendor libnetwork [#26879](https://github.com/docker/docker/pull/26879) [#26953](https://github.com/docker/docker/pull/26953)
+ * Avoid returning early on agent join failures [docker/libnetwork#1473](https://github.com/docker/libnetwork/pull/1473)
+
+### Swarm Mode
+
+* Do not autodetect advertise address on join [#26211](https://github.com/docker/docker/pull/26211)
+* Vendor SwarmKit [#26765](https://github.com/docker/docker/pull/26765)
+ * Bounce session after failed status update [docker/swarmkit#1539](https://github.com/docker/swarmkit/pull/1539)
+ * Fix possible deadlocks [docker/swarmkit#1537](https://github.com/docker/swarmkit/pull/1537)
```
Comment (aaronlehmann) [inline CHANGELOG.md]: "Fix possible raft deadlocks"

**Review by aaronlehmann — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 13 above)
Comment (vieux) [inline CHANGELOG.md]: done

File: CHANGELOG.md (hunk 1 above)
Comment (vieux) [inline CHANGELOG.md]: done

File: CHANGELOG.md (hunk 2 above)
Comment (vieux) [inline CHANGELOG.md]: done

File: CHANGELOG.md (hunk 3 above)
Comment (vieux) [inline CHANGELOG.md]: done

File: CHANGELOG.md (hunk 4 above)
Comment (vieux) [inline CHANGELOG.md]: done

File: CHANGELOG.md (hunk 5 above)
Comment (vieux) [inline CHANGELOG.md]: done

File: CHANGELOG.md (hunk 7 above)
Comment (vieux) [inline CHANGELOG.md]: done

File: CHANGELOG.md (hunk 8 above)
Comment (vieux) [inline CHANGELOG.md]: done

File: CHANGELOG.md (hunk 9 above)
Comment (vieux) [inline CHANGELOG.md]: done

File: CHANGELOG.md (hunk 10 above)
Comment (vieux) [inline CHANGELOG.md]: done

File: CHANGELOG.md (hunk 11 above)
Comment (vieux) [inline CHANGELOG.md]: removed

File: CHANGELOG.md (hunk 12 above)
Comment (vieux) [inline CHANGELOG.md]: done

File: CHANGELOG.md (hunk 17 above)
Comment (vieux) [inline CHANGELOG.md]: done

File: CHANGELOG.md [hunk 18]
```diff
@@ -5,6 +5,48 @@ information on the list of deprecated flags and APIs please have a look at
...
 be found.
 
+## 1.12.2-rc1 (2016-09-27)
+
+**IMPORTANT**: Docker 1.12 ships with an updated systemd unit file for rpm
+based installs (which includes RHEL, Fedora, CentOS, and Oracle Linux 7). When
+upgrading from an older version of docker, the upgrade process may not
+automatically install the updated version of the unit file, or fail to start
+the docker service if;
+
```
Comment (thaJeztah) [inline CHANGELOG.md]: Oh, there's more instructions below this (sorry, it's quite lengthy) https://github.com/docker/docker/blame/master/CHANGELOG.md#L10-L36

File: CHANGELOG.md (hunk 16 above)
Comment (vieux) [inline CHANGELOG.md]: ok I removed the line

File: CHANGELOG.md (hunk 15 above)
Comment (vieux) [inline CHANGELOG.md]: done

File: CHANGELOG.md (hunk 18 above)
Comment (vieux) [inline CHANGELOG.md]: fixed :)

**Review by thaJeztah — APPROVED** [review]
my earlier comments were addressed, so LGTM, but we may need more details for the Networking section

**Comment by vieux** [comment]
@aaronlehmann PTAL I addressed your comments

File: CHANGELOG.md [hunk 19]
```diff
@@ -5,6 +5,69 @@ information on the list of deprecated flags and APIs please have a look at
...
+- Fix exec's children handling [#26874](https://github.com/docker/docker/pull/26874)
+- Fix exec form of HEALTHCHECK CMD [#26208](https://github.com/docker/docker/pull/26208)
+
+### Networking
+* Vendor libnetwork [#26879](https://github.com/docker/docker/pull/26879) [#26953](https://github.com/docker/docker/pull/26953)
+ * Avoid returning early on agent join failures [docker/libnetwork#1473](https://github.com/docker/libnetwork/pull/1473)
+
+### Swarm Mode
+
+* Do not autodetect advertise address on join [#26211](https://github.com/docker/docker/pull/26211)
```
Comment (aaronlehmann) [inline CHANGELOG.md]: Sorry, just noticed this. I'd rephrase it as "Fix remote detection of a node's address when it joins the cluster".

Comment (vieux) [inline CHANGELOG.md]: no worries, I just pushed with this.

File: CHANGELOG.md [hunk 20]
```diff
@@ -5,6 +5,69 @@ information on the list of deprecated flags and APIs please have a look at
...
+### Runtime
+
+- Fix a panic due to a race condition filtering `docker ps` [#26049](https://github.com/docker/docker/pull/26049)
+* Implement retry logic to prevent "Unable to remove filesystem" errors when using the aufs storage driver [#26536](https://github.com/docker/docker/pull/26536)
+* Prevent devicemapper from removing device symlinks if `dm.use_deferred_removal` is enabled [#24740](https://github.com/docker/docker/pull/24740)
+- Fix an issue where the CLI did not return correct exit codes if a command was run with invalid options [#26777](https://github.com/docker/docker/pull/26777)
+- Fix a panic due to a bug in stdout / stderr processing in health checks [#26507](https://github.com/docker/docker/pull/26507)
+- Fix exec's children handling [#26874](https://github.com/docker/docker/pull/26874)
+- Fix exec form of HEALTHCHECK CMD [#26208](https://github.com/docker/docker/pull/26208)
+
```
Comment (aboch) [inline CHANGELOG.md]: Related to the boltdb vendoring, can you add the issue it fiexed here in the runtime section:

`Fix a daemon start panic on armv5 [#24315](https://github.com/docker/docker/issues/24315)`

Comment (vieux) [inline CHANGELOG.md]: added

Comment (aboch) [inline CHANGELOG.md]: It was a generic boltdb issue for armv5, triggered via networking code. If that is relevant, then we should add it in the Networking section.

**Comment by aaronlehmann** [comment]
LGTM

File: CHANGELOG.md (hunk 20 above)
Comment (vieux) [inline CHANGELOG.md]: done

File: CHANGELOG.md [hunk 21]
```diff
@@ -5,6 +5,71 @@ information on the list of deprecated flags and APIs please have a look at
...
+### Networking
+
+- Fix a daemon start panic on armv5 [#24315](https://github.com/docker/docker/issues/24315)
+* Vendor libnetwork [#26879](https://github.com/docker/docker/pull/26879) [#26953](https://github.com/docker/docker/pull/26953)
```
Comment (mrjana) [inline CHANGELOG.md]: * Fix service published port cleanup issues [docker/libetwork#1432](https://github.com/docker/libnetwork/pull/1432) [docker/libnetwork#1433](https://github.com/docker/libnetwork/pull/1433)
* Recover properly from transient gossip failures [docker/libnetwork#1446](https://github.com/docker/libnetwork/pull/1446)
* Disambiguate node names known to gossip cluster to avoid node name collision [docker/libnetwork#1451](https://github.com/docker/libnetwork/pull/1451)
* Honor user provided listen address for gossip  [docker/libnetwork#1460](https://github.com/docker/libnetwork/pull/1460)
* Allow reachability via published port across services on the same host [docker/libnetwork#1398](https://github.com/docker/libnetwork/pull/1398)
* Change the ingress sandbox name from random id to just `ingress_sbox` [docker/libnetwork#1449](https://github.com/docker/libnetwork/pull/1449)

Comment (vieux) [inline CHANGELOG.md]: added

Comment (mrjana) [inline CHANGELOG.md]: LGTM

**Review by mrjana — APPROVED** [review]
LGTM on networking change logs

**Review by thaJeztah — APPROVED** [review]
LGTM, looks good for the RC

**Comment by thaJeztah** [comment]
looks like @aaronlehmann left his LGTM as comment, so merging

**Comment by thaJeztah** [comment]
merged to master through https://github.com/docker/docker/pull/27328
